### PR TITLE
Feature/multi type params for mentions search

### PR DESCRIPTION
**Pull request by izzyblues**
As we agreed, we want that the search for mentions requires to specify the 4 concepts as types in the URL instead of empty concept type parameters

### CODE REVIEW

File: service/search_test.go [hunk 1]
```diff
@@ -222,34 +223,96 @@ func (s *EsConceptSearchServiceTestSuite) TestFindAllConceptsByTypeInvalid() {
...
+
+func (s *EsConceptSearchServiceTestSuite) TestSuggestConceptByTextAndTypeInvalidType() {
+	service := NewEsConceptSearchService(testIndexName, 10, 10, 2)
+	service.SetElasticClient(s.ec)
+
+	_, err := service.SuggestConceptByTextAndTypes("test", []string{"pippo"})
+	assert.EqualError(s.T(), err, fmt.Sprintf(errInvalidConceptTypeFormat, "pippo"), "expected invalid type error")
+}
+
+func (s *EsConceptSearchServiceTestSuite) TestSuggestConceptByTextAndTypesNotEnoghtValidTypes() {
```
Comment (khatton-ft) [inline service/search_test.go]: :pencil2: `Enoght`

**Comment by khatton-ft** [comment]
I do think this is misleading for clients; the query only works if they specify the 4 types. Isn't that going to lead to confusion? Why can't I just specify Location, for example?

File: service/search.go [hunk 2]
```diff
@@ -157,16 +197,48 @@ func (s *esConceptSearchService) SuggestAuthorsByText(textQuery string, conceptT
...
+	}
+	return nil
+}
+
+func (s *esConceptSearchService) SuggestConceptByTextAndTypesWithBoost(textQuery string, conceptTypes []string, boostType string) ([]Concept, error) {
 	if textQuery == "" {
-		return nil, ErrEmptyTextParameter
+		return nil, errEmptyTextParameter
+	}
+	if len(conceptTypes) == 0 {
```
Comment (peteclark-ft) [inline service/search.go]: Similarly to above, this should be checked in the handler.

File: api/api.yml [hunk 3]
```diff
@@ -34,7 +34,10 @@ paths:
             - http://www.ft.com/ontology/Genre
             - http://www.ft.com/ontology/product/Brand
             - http://www.ft.com/ontology/person/Person
-          required: false
+            - http://www.ft.com/ontology/organisation/Organisation
+            - http://www.ft.com/ontology/Location
+            - http://www.ft.com/ontology/Topic
```
Comment (izzyblues) [inline api/api.yml]: Yes, we did

**Comment by izzyblues** [comment]
@peteclark-ft I have the exact opposite vision on the code. I think that the `handler` should deal only with the consistency at the HTTP level, which means checking that all the requested parameters are in the URL. In my opinion, the `service` package should check the semantic consistency of the parameters value (e.g, non-empty and valid values). But If you think that this approach reduces the readability of the code, I can restore everything as it was. @khatton-ft Do you have any opinion on this?

File: api/api.yml [hunk 4]
```diff
@@ -28,13 +28,26 @@ paths:
       parameters:
         - name: type
           in: query
-          description: The type of Concept to search for as a URI. When used in isolation, the results will be the first 50 concepts of that type - this should only really be used with small collections such as Genres. When used in combination with other modes such as `mode=autocomplete`, this will restrict queries to search for concepts by the given type.
+          description: >
+            The type of Concept to search for as a URI.
+            When used in isolation, the results will be the first 50 concepts of
+            that type - this should only really be used with small collections
```
Comment (khatton-ft) [inline api/api.yml]: s/should only really be used with/is only useful for/

File: api/api.yml [hunk 5]
```diff
@@ -28,13 +28,26 @@ paths:
       parameters:
         - name: type
           in: query
-          description: The type of Concept to search for as a URI. When used in isolation, the results will be the first 50 concepts of that type - this should only really be used with small collections such as Genres. When used in combination with other modes such as `mode=autocomplete`, this will restrict queries to search for concepts by the given type.
+          description: >
+            The type of Concept to search for as a URI.
+            When used in isolation, the results will be the first 50 concepts of
```
Comment (khatton-ft) [inline api/api.yml]: s/in isolation, the/without a mode, only a single value for type can be used. The/

File: api/api.yml [hunk 6]
```diff
@@ -28,13 +28,26 @@ paths:
...
           type: string
           enum:
             - http://www.ft.com/ontology/Genre
             - http://www.ft.com/ontology/product/Brand
             - http://www.ft.com/ontology/person/Person
-          required: false
+            - http://www.ft.com/ontology/organisation/Organisation
+            - http://www.ft.com/ontology/Location
+            - http://www.ft.com/ontology/Topic
+          required:
```
Comment (khatton-ft) [inline api/api.yml]: is it or isn't it?

**Comment by khatton-ft** [comment]
IMO the handler has the responsibility for routing requests to the service and marshalling a response; this requires unpacking the HTTP request and checking for existence and cardinality of parameters to the extend necessary to make routing decisions, but not (beyond routing decisions) to examine or validate the values of each parameter. Does that help @izzyblues @peteclark-ft ?

**Comment by khatton-ft** [comment]
I'm concerned we are, effectively, requiring API clients to have deep knowledge about the structure of the underlying index. This is a consequence of being driven by a stream of narrow use-cases, but I do think we should be trying to keep the API generic.

We support
- a GET request by type, which returns up to 50 items of that type
- a GET request by type and autocomplete text, which returns up to 10 items of that type
- a GET request by a seemingly arbitrary but mandatory set of types and autocomplete text, which returns up to 10 items of that type

The second and third items in that list don't look appreciably different to an API client, but the internal behaviour is very different. I don't think this PR should be blocked because of it but I do think we need to consider that we've inadvertently accumulated some technical debt because this is not a straightforward API for a generic client to use.

**Comment by peteclark-ft** [comment]
@khatton-ft that's fair enough for point one. @izzyblues code looks much cleaner to me now 👍

**Comment by peteclark-ft** [comment]
@khatton-ft I completely agree re: API tech debt. It shouldn't block this PR though, as you say.

Maybe we should look at implementing fewer suggesters, which contain more categories, so that we can create more generic searches across the types in the index?

**Comment by khatton-ft** [comment]
@peteclark-ft As I understand it, the problem is that suggesters operate across the whole index and can't be used with filters, so it's a bit of a nightmare if you sometimes want to suggest for one type only, and at other times want to suggest across multiple types.

**Pull request merged**